File: apps/api/src/editor/dtos/snapshot.ts
export enum SnapshotState {
  Importing = "IMPORTING",
  Imported = "IMPORTED",
  Invalid = "INVALID",
}

export interface SnapshotDTO {
  readonly state: SnapshotState.Imported;
  readonly name: string;
  readonly path: string;
  readonly stats: SnapshotStatsDTO;
}

export interface SnapshotBeingImportedDTO {
  readonly state: SnapshotState.Importing;
  readonly name: string;
  readonly path: string;
}

export interface SnapshotStatsDTO {
  readonly fileSize: number;
  readonly createdAt: Date;
  readonly importedAt: Date;
}

export enum ImportSnapshotErrorCode {
  InvalidFile = "INVALID_FILE",
  Duplicate = "DUPLICATE",
}
export type ImportSnapshotValidationDTO =
  | { valid: true }
  | { valid: false; errorMessage: string; errorCode: ImportSnapshotErrorCode };

// Groups of instances categorized by their constructor.
export interface SummaryGroupDTO {
  constructor: string;
  distance: number;
  shallowSize: string;
  retainedSize: string;
  instanceCount: number;
}

File: apps/desktop-app/lib/src/api/dtos/snapshot.dart
enum SnapshotState { importing, imported, invalid }

class Snapshot {
  final String name;
  final String path;
  final int? fileSize;
  final SnapshotState state;
  final Stream<String>? progressStream;
  final String? errorMessage;

  Snapshot.importing(this.name, this.path, this.progressStream)
      : state = SnapshotState.importing,
        fileSize = null,
        errorMessage = null;

  Snapshot.invalid(this.name, this.path, this.errorMessage)
      : state = SnapshotState.invalid,
        progressStream = null,
        fileSize = null;

  Snapshot(this.name, this.path, this.fileSize)
      : state = SnapshotState.imported,
        progressStream = null,
        errorMessage = null;
}
